jgrapht-core/src/main/java/org/jgrapht/alg/tour/FarthestInsertionHeuristicTSP.java

```diff
@@ -0,0 +1,319 @@
+/*
+ * (C) Copyright 2019-2021, by Peter Harman and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * See the CONTRIBUTORS.md file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the
+ * GNU Lesser General Public License v2.1 or later
+ * which is available at
+ * http://www.gnu.org/licenses/old-licenses/lgpl-2.1-standalone.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR LGPL-2.1-or-later
+ */
+package org.jgrapht.alg.tour;
+
+import org.jgrapht.Graph;
+import org.jgrapht.GraphPath;
+import org.jgrapht.Graphs;
+import org.jgrapht.graph.GraphWalk;
+import org.jgrapht.util.ArrayUtil;
+import org.jgrapht.util.VertexToIntegerMapping;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * The farthest insertion heuristic algorithm for the TSP problem.
+ *
+ * <p>
+ * The travelling salesman problem (TSP) asks the following question: "Given a list of cities and
+ * the distances between each pair of cities, what is the shortest possible route that visits each
+ * city exactly once and returns to the origin city?".
+ * </p>
+ *
+ * <p>
+ * Insertion heuristics are quite straightforward, and there are many variants to choose from. The
+ * basics of insertion heuristics is to start with a partial tour of a subset of all cities,
+ * and then iteratively an unvisited vertex (a vertex whichis not in the tour) is chosen given a criterion
+ * and inserted in the best position of the partial tour. Per each iteration, the farthest insertion
+ * heuristic selects the farthest unvisited vertex from the partial tour.
+ * This algorithm provides a guarantee to compute tours no more than
+ * 0(log N) times optimum (assuming the triangle inequality). However, regarding practical results,
+ * some references refer to this heuristic as one of the best among the category of insertion heuristics.
+ * This implementation uses the longest edge by default as the initial sub-tour if one is not provided.
+ * </p>
+ *
+ * <p>
+ * The description of this algorithm can be consulted on: <br>
+ * Johnson, D. S., & McGeoch, L. A. (2007). Experimental Analysis of Heuristics for the STSP.
+ * In G. Gutin & A. P. Punnen (Eds.), The Traveling Salesman Problem and Its Variations (pp. 369–443).
+ * Springer US. https://doi.org/10.1007/0-306-48213-4_9
+ * </p>
+ *
+ * <p>
+ * This implementation can also be used in order to augment an existing partial tour. See
+ * constructor {@link #FarthestInsertionHeuristicTSP(GraphPath)}.
+ * </p>
+ *
+ * <p>
+ * The runtime complexity of this class is $O(V^2)$.
+ * </p>
+ *
+ * <p>
+ * This algorithm requires that the graph is complete.
+ * </p>
+ *
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ * @author José Alejandro Cornejo Acosta
+ */
```

Comment:
```suggestion
 * @param <V> the graph vertex type
 * @param <E> the graph edge type
 * @author José Alejandro Cornejo Acosta
 * 
 * @since 1.5.3
 */
```